src/software/new_geom/polynomial.h

```diff
@@ -0,0 +1,156 @@
+#pragma once
+
+#include <cmath>
+#include <vector>
+
+/**
+ * Polynomial is a representation of a polynomial
+ * that can calculate values at a given input
+ */
+class Polynomial
+{
+   public:
+    // Threshold for coefficient comparison.
+    // http://www.cplusplus.com/forum/beginner/95128/
+    static constexpr double EPSILON = 1e-15;
+
+    /**
+     * Construct a zero polynomial
+     */
+    explicit Polynomial();
+
+    /**
+     * Construct a polynomial from coefficients of the form
+     * coeffs[0] + coeffs[1]*x^(1) + ... + coeffs[n-1]*x^(n-1)
+     *
+     * @param coeffs coefficients of the polynomial
+     */
+    explicit Polynomial(const std::vector<double> &coeffs);
+
+    /**
+     * Construct a polynomial from coefficients of the form
+     * coeffs[0] + coeffs[1]*x^(1) + ... + coeffs[n-1]*x^(n-1)
+     *
+     * @param coeffs coefficients of the polynomial
+     */
+    explicit Polynomial(const std::initializer_list<double> &coeffs);
+
+    /**
+     * Construct a linear polynomial from two pairs of input/output
+     *
+     * @param constraint1 first pair of values: first is the input, second is the output
+     * @param constraint2 second pair of values: first is the input, second is the output
+     *
+     * @throws std::invalid_argument if constraint1.first == constraint2.first
+     */
+    explicit Polynomial(const std::pair<double, double> &constraint1,
+                        const std::pair<double, double> &constraint2);
+
+    /**
+     * Returns the coefficient of the term of given order
+     *
+     * @param order the order of the term
+     * @return the coefficient of the term
+     */
+    double getCoeff(unsigned int order) const;
+
+    /**
+     * Sets the coefficient of the term of given order
+     *
+     * @param order the order of the term to set the coefficient
+     * @param coeff the coefficient
+     */
+    void setCoeff(unsigned int order, double coeff);
+
+    /**
+     * Returns the order of the Polynomial
+     *
+     * @return the order of the Polynomial
```

Comment:
``` suggestion
     * @return the index-1 of the highest order non-zero coefficient
```

Reply:
Done in af20cdc76f49a87c92abe851d79b90511c7bfb33, although I believe it should be `index` not `index-1`

Reply:
Doesn't look like this was changed?

Reply:
My mistake, edited the wrong comment. Fixed here: 1a85e1b1017c773acfad28be47a5f425a06afa67

Reply:
My mistake, edited the wrong comment. Fixed here: 1a85e1b